proposed/2021/PackageNamespaces.md

```diff
@@ -0,0 +1,568 @@
+
+    <!-- Add a namespace for Contoso.* under the contoso source. -->
+    <packageSource name="contoso">
+        <namespace id="Contoso.*" />
+        <!-- Add a specific id to be download from  the contoso source. -->
+        <namespace id="Special.Package.With.An.Imperfect.Id" />
+    </packageSource>
+ 
+</packageNamespaces>
+```
+
+By default, namespaces will allow a package to match multiple namespaces and download with no precedence. In some cases, one might want to avoid ambiguous cases. That can be done by adding a `strict` flag.
+
+```xml
+<config>
+    <add key="namespaceMode" value="strict" />
+</config>
+```
+
+The strict flag provides a source pinning behavior by nature in which one namespace is allowed per package source. Entering a strict mode will provide users an error experience when there are namespace conflicts to which a user will be able to resolve by defining one namespace per package source configuration.
+
+An error for this experience might look like:
+
+```console
+NUXXXX: Package namespace {0} is listed on the following sources: {1}. Only one unique package namespace can be defined across sources.
+```
+
+### Technical explanation
+
+<!-- Explain the proposal in sufficient detail with implementation details, interaction models, and clarification of corner cases. -->
+
+#### Technical background - how NuGet package installation works today
+
+This approach builds on top of the current package installation behavior as described in the [package installation process](https://docs.microsoft.com/nuget/concepts/package-installation-process).
+
+A few other important concepts:
+
+- global packages folder - The installation directory for all packages in PackageReference. Think of this as the `Program Files` equivalent. All packages are consumed from this location. A single package installation can be shared safely among many different builds. A project *does not own* a specific package installation.
+The global packages folder is an *append only* resource. This means NuGet only ever installs a new package. There is **no** refreshing, or overriding, or `re-installing` packages of any kind.
+- When requesting an exact version, local, file sources are currently preferred - When installing a package, NuGet checks local sources independently before checking http sources.
+- Package installation is operation based - If 3 projects are being restored during that operation, and all those projects have a dependency to `Newtonsoft.Json`, version `9.0.1`, in regular scenarios, only 1 project will *actually* download and install the package. The other projects will use the already installed version.
+- `.nupkg.metadata` - Each package installation directory contains a `.nupkg.metadata` file which signify that a package installation is complete. This is expected to be the last file written in the package directory. This file is *not* use during build. NuGet.Client writes the package source information inside this file.
```

Comment:
Do earlier versions of NuGet clients already write the package source information to `.nuget.metadata`? What will the new clients do if the file exists but does not contain any source information?

Reply:
@nkolev92 May know better. AFAIK earlier versions of client will ignore these new elements & possibly rewrite them as it won't be recognized. New clients would know & understand it.

Reply:
Once a package has been installed in the global packages folder, no changes will happen from NuGet side. 
So the file will remain as is. 

For the time being, the information in the nupkg metadata is for auditing purposes only, NuGet tooling writes it, but is not using it yet.